### PR TITLE
✨ Add webhook to check if same serverID baremetal host exist or not

### LINKED ISSUE #928: Webhook testing that not multiple HetznerBareMetalHost objects share the same serverID
## Description

A webhook that tests for HetznerBareMetalHosts on creation whether other hosts exist that have the same ServerID in the spec. If so, we should return an error that this must not happen.

If the client.List call of objects fails, we should emit a warning that we could not verify that the host has a unique serverID

### PR DESCRIPTION
**Pull request by aniruddha2000**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #928 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squash commits
- [ ] include documentation
- [ ] add unit tests

### CODE REVIEW

File: api/v1beta1/hetznerbaremetalhost_webhook.go [hunk 1]
```diff
@@ -40,19 +55,41 @@ func (host *HetznerBareMetalHost) Default() {
...
+	hetznerBareMetalHostObj, ok := (obj).(*HetznerBareMetalHost)
+	if !ok {
+		return admission.Warnings{}, errors.NewBadRequest(fmt.Sprintf("expected HetznerBareMetalHost, but got %T", hetznerBareMetalHostObj))
+	}
+
+	var allErrs field.ErrorList
+
+	hetznerBareMetalHostList := &HetznerBareMetalHostList{}
+	if err := host.client.List(ctx, hetznerBareMetalHostList, &ctrlClient.ListOptions{Namespace: hetznerBareMetalHostObj.Namespace}); err != nil {
+		allErrs = append(allErrs,
```
Comment (janiskemper) [inline api/v1beta1/hetznerbaremetalhost_webhook.go:71]: not an error, a warning.

File: api/v1beta1/hetznerbaremetalhost_webhook.go [hunk 2]
```diff
@@ -40,19 +55,41 @@ func (host *HetznerBareMetalHost) Default() {
...
+			field.InternalError(field.NewPath("spec", "ServerID"), fmt.Errorf("could not verify that the host has a unique serverID")),
+		)
+	}
+
+	for _, hetznerBareMetalHost := range hetznerBareMetalHostList.Items {
+		if hetznerBareMetalHost.Spec.ServerID == hetznerBareMetalHostObj.Spec.ServerID {
+			allErrs = append(allErrs,
+				field.Invalid(field.NewPath("spec", "ServerID"), hetznerBareMetalHostObj.Spec.ServerID, fmt.Sprintf("other host exist with same ServerID: %d", hetznerBareMetalHostObj.Spec.ServerID)),
```
Comment (janiskemper) [inline api/v1beta1/hetznerbaremetalhost_webhook.go:79]: maybe mention not the serverID (that's obvious anyway), maybe mention the name of the other host object with the same ServerID

File: api/v1beta1/hetznerbaremetalhost_webhook.go [hunk 3]
```diff
@@ -40,19 +55,41 @@ func (host *HetznerBareMetalHost) Default() {
 
 //+kubebuilder:webhook:path=/validate-infrastructure-cluster-x-k8s-io-v1beta1-hetznerbaremetalhost,mutating=false,failurePolicy=fail,sideEffects=None,groups=infrastructure.cluster.x-k8s.io,resources=hetznerbaremetalhosts,verbs=create;update,versions=v1beta1,name=validation.hetznerbaremetalhost.infrastructure.cluster.x-k8s.io,admissionReviewVersions={v1,v1beta1}
 
-var _ webhook.Validator = &HetznerBareMetalHost{}
+var _ webhook.CustomValidator = &HetznerBareMetalHostWebhook{}
 
 // ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
-func (host *HetznerBareMetalHost) ValidateCreate() (admission.Warnings, error) {
-	return nil, nil
+func (host *HetznerBareMetalHostWebhook) ValidateCreate(ctx context.Context, obj runtime.Object) (admission.Warnings, error) {
```
Comment (janiskemper) [inline api/v1beta1/hetznerbaremetalhost_webhook.go:61]: maybe don't name this host but something else (hw for host webhook maybe?) and instead the hetznerBareMetalHostObj should be host

**Review by janiskemper — CHANGES_REQUESTED** [review]

File: api/v1beta1/hetznerbaremetalhost_webhook.go (hunk 2 above)
Comment (aniruddha2000) [reply api/v1beta1/hetznerbaremetalhost_webhook.go:79]:
```go
return admission.Warnings{"could not verify that the host has a unique serverID"}, nil
```

Lint failure - `error is not nil (line 70) but it returns nil (nilerr)`

File: api/v1beta1/hetznerbaremetalhost_webhook.go (hunk 1 above)
Comment (aniruddha2000) [reply api/v1beta1/hetznerbaremetalhost_webhook.go:71]:
```go
return admission.Warnings{"could not verify that the host has a unique serverID"}, nil
```

Lint failure - `error is not nil (line 70) but it returns nil (nilerr)`

Comment (janiskemper) [reply api/v1beta1/hetznerbaremetalhost_webhook.go]: you can still put the error message in the warning! That might even be very useful!

**Review by janiskemper — DISMISSED** [review]

File: api/v1beta1/hetznerbaremetalhost_webhook.go [hunk 4]
```diff
@@ -43,28 +57,57 @@ func (host *HetznerBareMetalHost) Default() {
...
+
+	hetznerBareMetalHostList := &HetznerBareMetalHostList{}
+	if err := hw.client.List(ctx, hetznerBareMetalHostList, &ctrlClient.ListOptions{Namespace: host.Namespace}); err != nil {
+		return admission.Warnings{"could not verify that the host has a unique serverID"}, fmt.Errorf("failed to list hetznerBareMetalHost: %w", err)
+	}
+
+	for _, hetznerBareMetalHost := range hetznerBareMetalHostList.Items {
+		if hetznerBareMetalHost.Spec.ServerID == host.Spec.ServerID {
+			allErrs = append(allErrs,
+				field.Invalid(field.NewPath("spec", "ServerID"), host.Spec.ServerID, fmt.Sprintf("%q host exist with same ServerID: %d", hetznerBareMetalHost.Name, host.Spec.ServerID)),
```
Comment (janiskemper) [inline api/v1beta1/hetznerbaremetalhost_webhook.go:79]: serverID not ServerID

**Review by janiskemper — CHANGES_REQUESTED** [review]

File: api/v1beta1/hetznerbaremetalhost_webhook.go [hunk 5]
```diff
@@ -43,28 +57,57 @@ func (host *HetznerBareMetalHost) Default() {
...
+func (hw *HetznerBareMetalHostWebhook) ValidateCreate(ctx context.Context, obj runtime.Object) (admission.Warnings, error) {
+	host, ok := (obj).(*HetznerBareMetalHost)
+	if !ok {
+		return admission.Warnings{}, errors.NewBadRequest(fmt.Sprintf("expected HetznerBareMetalHost, but got %T", host))
+	}
+
+	var allErrs field.ErrorList
+
+	hetznerBareMetalHostList := &HetznerBareMetalHostList{}
+	if err := hw.client.List(ctx, hetznerBareMetalHostList, &ctrlClient.ListOptions{Namespace: host.Namespace}); err != nil {
```
Comment (janiskemper) [inline api/v1beta1/hetznerbaremetalhost_webhook.go:72]: no namespace scope

File: api/v1beta1/hetznerbaremetalhost_webhook.go [hunk 6]
```diff
@@ -43,28 +57,57 @@ func (host *HetznerBareMetalHost) Default() {
...
+		return admission.Warnings{"could not verify that the host has a unique serverID"}, fmt.Errorf("failed to list hetznerBareMetalHost: %w", err)
```
Comment (janiskemper) [inline api/v1beta1/hetznerbaremetalhost_webhook.go:73]: I meant to integrate the error in the warning message, not returning an error :see_no_evil:

File: api/v1beta1/hetznerbaremetalhost_webhook.go [hunk 7]
```diff
@@ -17,19 +17,33 @@ limitations under the License.
...
 	ctrl "sigs.k8s.io/controller-runtime"
+	ctrlClient "sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/webhook"
 	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
 )
 
+// HetznerBareMetalHostWebhook implements validating and defaulting webhook for HetznerBareMetalHost.
+// +k8s:deepcopy-gen=false
+type HetznerBareMetalHostWebhook struct {
+	client ctrlClient.Client
```
Comment (janiskemper) [inline api/v1beta1/hetznerbaremetalhost_webhook.go:37]: make it c client.Client
I think that's more aligned with the rest of the code

**Review by janiskemper — CHANGES_REQUESTED** [review]

File: api/v1beta1/hetznerbaremetalhost_webhook.go [hunk 8]
```diff
@@ -34,12 +33,12 @@ import (
 // HetznerBareMetalHostWebhook implements validating and defaulting webhook for HetznerBareMetalHost.
 // +k8s:deepcopy-gen=false
 type HetznerBareMetalHostWebhook struct {
-	client ctrlClient.Client
+	c ctrlClient.Client
```
Comment (janiskemper) [inline api/v1beta1/hetznerbaremetalhost_webhook.go:36]: client.Client not ctrlClient

File: api/v1beta1/hetznerbaremetalhost_webhook.go [hunk 9]
```diff
@@ -43,28 +56,57 @@ func (host *HetznerBareMetalHost) Default() {
...
+	newHost, ok := newObj.(*HetznerBareMetalHost)
+	if !ok {
+		return admission.Warnings{}, apierrors.NewBadRequest(fmt.Sprintf("expected an ClusterStack but got a %T", newHost))
+	}
+
 	var allErrs field.ErrorList
 
-	oldHetznerBareMetalHost := old.(*HetznerBareMetalHost)
-	if !reflect.DeepEqual(host.Spec.ServerID, oldHetznerBareMetalHost.Spec.ServerID) {
+	if !reflect.DeepEqual(newHost.Spec.ServerID, oldHost.Spec.ServerID) {
```
Comment (janiskemper) [inline api/v1beta1/hetznerbaremetalhost_webhook.go:100]: we don't need deepequal right? Just normal comparison

**Review by janiskemper — DISMISSED** [review]

File: pkg/services/baremetal/host/host_suite_test.go [hunk 10]
```diff
@@ -36,7 +36,7 @@ import (
 )
 
 const (
-	bareMetalHostID  = 1
+	// bareMetalHostID  = 1
```
Comment (janiskemper) [inline pkg/services/baremetal/host/host_suite_test.go:39]: I guess you can remove this entirely, right?

**Review by janiskemper — DISMISSED** [review]
looks good. You can maybe remove this one commented out line

**Review by janiskemper — APPROVED** [review]

**Pull request merged**